### ISSUE TITLE
Official Helm Chart

### ISSUE THREAD

**Issue by stevehipwell** [kubernetes-sigs/descheduler#248]
Now that there is an official Docker image (#144) it would be really useful if there was an official Helm chart.

**Comment by seanmalloy** [comment]
Here are the required details that are needed to make this issue qualify as a good first issue.

Please use helm v3, not v2.

The helm chart needs to create a `CronJob`, `ConfigMap`, `ClusterRole`, `ServiceAccount`, and `ClusterRoleBinding` in the `kube-system` namespace. The `.spec.schedule` field needs to be configurable. The `ConfigMap` needs to support specifying a custom descheduler policy. See the files in the [kubernetes directory](https://github.com/kubernetes-sigs/descheduler/blob/master/kubernetes/rbac.yaml) for a good starting point.

The descheduler container image can be pulled from the below registry locations.

`asia.gcr.io/k8s-artifacts-prod/descheduler/descheduler:v0.10.0`
`eu.gcr.io/k8s-artifacts-prod/descheduler/descheduler:v0.10.0`
`us.gcr.io/k8s-artifacts-prod/descheduler/descheduler:v0.10.0`

**Comment by stevehipwell** [comment]
@seanmalloy I'm interested in picking this up and am slightly curious around the Helm v3 requirement? I understand that in a perfect world we'd all have moved to Helm v3 and all of the chart authors would have upgraded their charts to v3. However based on the current available charts it seems that most people are still using v2 charts even if they are applying them via v3. My suggestion to add this chart was to get closer alignment and it seems like only supporting Helm v3 might not be the best way to do this?

Would it be worth creating two charts (v2 & v3) or due to the limited scope of this chart would it be better sticking with Helm v2 for now?

Also do you have a repository for this new chart as the [stable repo is no longer taking new chart submissions](https://github.com/helm/charts#deprecation-timeline)?

**Comment by Odernor** [comment]
Hi,
i have written an small descheduler helm chart for me. Im newbie in this..
Perhaps it can help an little: https://github.com/Odernor/descheduler_helm
its helm v2 but is installable with v3 too

**Comment by stevehipwell** [comment]
The following chart has been around for a while as the author also created a Docker image.
https://github.com/komljen/helm-charts/tree/master/descheduler

**Comment by seanmalloy** [comment]
> I'm interested in picking this up and am slightly curious around the Helm v3 requirement?

I'm fine with Helm v2 if you think that makes the most sense.

> Also do you have a repository for this new chart as the stable repo is no longer taking new chart submissions?

I'm not aware of an official location for storing k8s project helm charts. I'll reach out to [wg-k8s-infra](https://github.com/kubernetes/community/tree/master/wg-k8s-infra) on slack to see if there is an official k8s helm repo.

Other ideas ...
We could use GitHub pages in this repo as a helm repo. I've set this up for one of the other projects I maintain. There was a GCS bucket setup for descheduler when GCR was setup. Maybe the GCS bucket could be used as helm repo. I believe the staging GCS bucket for descheduler is `gs://k8s-staging-descheduler`.

**Comment by seanmalloy** [comment]
> I'm interested in picking this up

@stevehipwell feel free to start working on this. Let me know if you have any additional questions.

**Comment by seanmalloy** [comment]
@stevehipwell I've asked in the `#wg-k8s-infra` slack channel to see if the k8s project has an official helm repo that can be used. I'll let you know once I get an answer.

**Comment by seanmalloy** [comment]
I did not get any responses in the `#wg-k8s-infra` slack channel. My suggestion is that we use [GitHub Pages](https://pages.github.com/) in this repo to host the helm repo. See [this link](https://github.com/KohlsTechnology/eunomia/tree/gh-pages) for an example of what this would look like.

@ravisantoshgudimetla @aveshagarwal @damemi any concerns with using GitHub pages to host the helm repo for descheduler?

**Comment by stevehipwell** [comment]
@seanmalloy I've been on holiday for the last week but will pick this up in the next couple of days. I'm happy with the GitHub pages repo suggestion and would look at using GitHub Actions to achieve this as I've used them successfully before.

**Comment by ndegory** [comment]
@seanmalloy , @stevehipwell  any reason not to submit a new chart in https://github.com/helm/charts? (probably incubator). I'm using the chart mentioned above, but it's out of date, and PR are waiting for consideration, if that can help speed up to initiate the official chart https://github.com/komljen/helm-charts/pull/30

**Comment by stevehipwell** [comment]
@ndegory see previous comment.

> Also do you have a repository for this new chart as the [stable repo is no longer taking new chart submissions](https://github.com/helm/charts#deprecation-timeline)?

**Comment by ndegory** [comment]
@stevehipwell did you have time to work on it?

**Comment by stevehipwell** [comment]
@ndegory I've not had a chance yet but I'm hoping to find some time this week. Once I have the time it should be very quick to get something done.

**Comment by stevehipwell** [comment]
@seanmalloy how does compatibility between releases work? Is it the same as usual with v0.17 being compatible with K8s v1.16, v1.17 and v1.18?

**Comment by seanmalloy** [comment]
> @seanmalloy how does compatibility between releases work? Is it the same as usual with v0.17 being compatible with K8s v1.16, v1.17 and v1.18?

@stevehipwell yes that is correct for now. See #288.

**Comment by stevehipwell** [comment]
Thanks @seanmalloy.

RE Helm, I'm thinking a top level Helm directory containing the chart?

**Comment by seanmalloy** [comment]
> RE Helm, I'm thinking a top level Helm directory containing the chart?

That sounds good to me.

**Comment by stevehipwell** [comment]
@seanmalloy I've opened #298 which is close to being finished. I need to know what the release process for the descheduler is so I can try and align the release GitHub action to that, assuming this is what's wanted. There will also need to be a token added to the config for the release to use (I'll add instructions to the PR when ready). Finally someone might want to update the chart docs with a better overview before actually creating a release.

**Comment by ThoTischner** [comment]
Hey just check out our helm chart, its simple: https://bitsbeats.github.io/helm-charts/

**Comment by damemi** [comment]
> Hey just check out our helm chart, its super simple but made in germany: https://bitsbeats.github.io/helm-charts/

@ThoTischner thanks for the example! We were actually looking for examples of how other people are publishing their charts, would you mind taking a look at some of the discussion in https://github.com/kubernetes-sigs/descheduler/pull/298 (it's a bit long, warning) and give your insight?

It looks like we're going to go with publishing the chart on a separate tagged branch, which will automatically trigger publishing through Github actions when PRs are merged to it. Does that sound similar to your implementation?

**Comment by ThoTischner** [comment]
@damemi I had a look at the PR. So it looks good to me. I did not know the chart- tagging logic. It seems to be a good idea if you have no separate repo for your charts, like we have. Our chart repo is based on the chart repo from velero: https://github.com/vmware-tanzu/helm-charts 
Maybe it is a good idea to have separate repos for your app / code and the helm chart to keep things easy and do not mix up to many stuff in one big repo ;)

**Comment by damemi** [comment]
@ThoTischner I appreciate you taking the time to look at our PR, having the input from several different sources has been very helpful.

I agree that a separate repo might be the cleanest way, though I don't know if it's feasible for us to request an entirely new repo as a sig project. So, based on your feedback I think the branching implementation in the PR is a good compromise on that level. It keeps the charts "separate" from the project code while still being reasonably managed in the same space.

Thanks for your feedback!

**Comment by seanmalloy** [comment]
Thanks to @stevehipwell the descheduler helm chart is merged into the master branch!

I've opened the below two pull requests to wrap up the last few steps before releasing the helm chart.
* https://github.com/kubernetes-sigs/descheduler/pull/352
* https://github.com/kubernetes-sigs/descheduler/pull/351

**Comment by seanmalloy** [comment]
All of the code is merged into the `master` and `release-1.18` branches. I've asked @damemi to create the `chart-0.18.0` tag from the `release-1.18` branch.

/remove-good-first-issue

**Comment by damemi** [comment]
@seanmalloy created the release/tag at https://github.com/kubernetes-sigs/descheduler/releases/tag/chart-0.18.0, let's see if that kicks off the Helm-releaser action

**Comment by seanmalloy** [comment]
@damemi just creating a tag or release will trigger the GitHub Action. Looks like helm chart release automation ran.

https://github.com/kubernetes-sigs/descheduler/runs/894485507?check_suite_focus=true

**Comment by seanmalloy** [comment]
Seems like it did not work like we hoped, `Nothing to do. No chart changes detected`.

<img width="436" alt="Screen Shot 2020-07-21 at 9 04 48 AM" src="https://user-images.githubusercontent.com/685221/88064715-433b0980-cb31-11ea-93e8-ad58766ad4d7.png">

**Comment by damemi** [comment]
@seanmalloy beat me to it, I was going to point that out. Maybe we need to actually push a change to the tag. I also don't see the `gh-pages` branch available as a source for github pages

![Screen Shot 2020-07-21 at 10 07 29 AM](https://user-images.githubusercontent.com/1839101/88064999-06274500-cb3a-11ea-8acd-688908e5cc4e.png)

**Comment by seanmalloy** [comment]
@damemi I'm not familiar with this particular GitHub action. But my best guess would be to create a `gh-pages` branch and enable GitHub pages using this newly created branch.

**Comment by damemi** [comment]
@stevehipwell said the action should create the pages branch (https://github.com/kubernetes-sigs/descheduler/pull/298#issuecomment-659547978), would like to wait for his input as he may be able to answer our questions here :)

**Comment by stevehipwell** [comment]
Sorry @seanmalloy & @damemi I didn't check the docs properly and the releaser doesn't create the _gh-pages_ branch (yet). My suggestion would be to manually create an empty branch and then re-run the action.

**Comment by damemi** [comment]
Thanks @stevehipwell, I created the gh-pages branch and that source looks good to enable pages.

What about the `Discovering changed charts since 'chart-0.18.0'... Nothing to do. No chart changes detected.` message in the release action? What we did was merge the changes to the branch, then create the tag from the branch, so it sounds like the releaser action doesn't pick up changes from just a create. Should we do it the other way, create the branch/tag *then* push our version update to it? I just re-ran the action and got the same message

If that's the case, then we would have to push some small change to the branch. Like a `v0.18.1` change, that sound good @seanmalloy ?

**Comment by damemi** [comment]
> @damemi have you tried the action again since you created the branch? It's meant to check the version in Chart.yaml with the metadata on _gh-pages_.

@stevehipwell I did, and again just now, same output:
```
Run helm/chart-releaser-action@v1.0.0-rc.2
  with:
    charts_dir: charts
  env:
    CR_TOKEN: ***
descheduler
Looking up latest tag...
Discovering changed charts since 'chart-0.18.0'...
Nothing to do. No chart changes detected.
```

What I'm thinking is if we merged the changes, then created the chart-* tag, then created the github pages from that, then there won't be any differences between the chart tag and the gh-pages branch. So I think we need to push a small change to the chart- tag so that it can differ from gh-pages

**Comment by damemi** [comment]
I created https://github.com/kubernetes-sigs/descheduler/pull/354 to bump the chart to 0.18.1. When that merges, I'll create a new `chart-0.18.1` tag which should differ from the (now) existing `gh-pages` branch.

**Comment by stevehipwell** [comment]
@damemi I think you're probably right. It might be cleaner to remove the tag, make a minor change to get ahead of the _gh-pages_ branch and then re add the tag. That or update Chart.yaml to `0.18.1` and tag that change as `chart-0.18.1`.

**Comment by seanmalloy** [comment]
@damemi I think you might want to get rid of all the "extra" files on the `gh-pages` branch. Literally delete all the files from the `gh-pages` branch.

**Comment by seanmalloy** [comment]
The `gh-pages` branch will eventually need to look like something like this ... https://github.com/KohlsTechnology/eunomia/tree/gh-pages

**Comment by damemi** [comment]
> @damemi I think you might want to get rid of all the "extra" files on the `gh-pages` branch. Literally delete all the files from the `gh-pages` branch.

@seanmalloy do you think that's necessary or just for cleaning up? (since I don't think we were planning to keep gh-pages maintained for anything besides the chart right now) If it's just looking at differences in the charts I think the version bump should trigger it

**Comment by seanmalloy** [comment]
@damemi you can try with out doing the cleanup. All the gh-pages branch does is act like a webserver

**Comment by damemi** [comment]
Pushing a change didn't do it either, same "Nothing to do" message...

I see that message comes from after calling [`lookup_changed_charts`](https://github.com/helm/chart-releaser-action/blob/master/cr.sh#L58-L81) in the action. When I run that [`diff` command](https://github.com/helm/chart-releaser-action/blob/master/cr.sh#L200) locally from the `gh-pages` branch with the latest commit sha from `chart-0.18.1`, I do see the diff:

```
[(HEAD detached at upstream/gh-pages)] ~/go/src/sigs.k8s.io/descheduler$ git diff --find-renames --name-only c6c3be6362d8b7cbbe525a5f3af4c1401fa07378 -- charts/
charts/descheduler/Chart.yaml
```

so this should pass the [check](https://github.com/helm/chart-releaser-action/blob/master/cr.sh#L60) right?

**Comment by damemi** [comment]
Maybe I should have made the `gh-pages` branch from `release-1.18` instead of `master`?

**Comment by stevehipwell** [comment]
@damemi the _gh-pages_ branch should be empty, it might help is it was created from an earlier commit. I don't remember any of these issues on my repo, so I'm not sure what the difference is. It might be due to the cherry picking I guess.

**Comment by damemi** [comment]
Tried in [my fork](https://github.com/damemi/descheduler/runs/895931536?check_suite_focus=true) with the `gh-pages` branch empty and got the same result.

Going to try in my fork now with a `gh-pages` branch created from `master`, and add a new tag from `master` (ie, no cherry-picking, simulating for example the next, 1.19, release)

**Comment by damemi** [comment]
Update: creating a gh-pages branch from master (and emptying it) also doesn't detect any changes, even when I push a new  commit/tag after.

**Comment by damemi** [comment]
For reference, I tried this tag: https://github.com/damemi/descheduler/tree/chart-0.19.99

against this empty gh-pages branch: https://github.com/damemi/descheduler/tree/gh-pages

gh-pages is 1 commit behind the chart-0.19.99 tag: https://github.com/damemi/descheduler/compare/gh-pages...chart-0.19.99

Still nothing:
```
Run helm/chart-releaser-action@v1.0.0-rc.2
descheduler
Looking up latest tag...
Discovering changed charts since 'chart-0.19.99'...
Nothing to do. No chart changes detected.
```

Giving up on this for now, but if there are more ideas we can give them a shot

**Comment by stevehipwell** [comment]
@damemi I'll take a look on my repo tomorrow when I'm at a computer to see if there are any differences I can spot.

**Comment by damemi** [comment]
Another possibility: the `actions/checkout` action that runs before chart-releaser by default[ checks out the ref that triggered the workflow](https://github.com/actions/checkout#checkout-v2). In this case, that would mean it's checking out the `chart-*` tag, which the chart-releaser flow is comparing to itself.

So maybe we want to configure the checkout action (https://github.com/kubernetes-sigs/descheduler/blob/master/.github/workflows/release.yaml#L13) to have [`ref` point to a different branch](https://github.com/actions/checkout#usage), either `master` or `gh-pages`

**Comment by stevehipwell** [comment]
@damemi it looks like the assumption that I made over swapping out the default trigger of push to master to pushing a tag was incorrect as the action is hardcoded to compare the current ref to the latest tag. I'll try to think of an alternative and get back to you.

**Comment by damemi** [comment]
I got it to work in my own repo by switching the `push: tags` to push on `master` and `release-*` branches. This successfully created a tag and published a release of the helm chart (the tag and release are both `descheduler-<version>` (changes in https://github.com/kubernetes-sigs/descheduler/pull/356). (see successful runs to [`release-1.18`](https://github.com/damemi/descheduler/runs/899682590?check_suite_focus=true) and [`master`](https://github.com/damemi/descheduler/runs/899660418?check_suite_focus=true)

Reading more into the helm releaser tool, the tool itself is what creates the new tag/release when publishing the chart. So we shouldn't be trying to create/push the tag (from what I can tell this is inherently how the `cr` tool works).

Downsides to this: 
* I don't see any way we can change what the tag name or release title is. That means they'll always be `descheduler-<version>` (see https://github.com/damemi/descheduler/releases), which will get confusing against our regular releases. Maybe if we change the repo name in the chart to something like `descheduler-helm`
* We have a broken tag: https://github.com/kubernetes-sigs/descheduler/tree/0.9.0. This tag is missing the leading `v`, so it doesn't show up in order with the others. In fact, it shows as the "latest" tag, and if you look at my release runs I linked above you'll see that they always compare to this as the "latest". Imo, it is a bug in the chart-releaser-action to assume that the "latest" (alphabetically) tag is what should be compared against, and it doesn't have any logic to actually check the latest chart release tag. I don't think we can just delete/rename this tag since that could seriously break anyone that is currently depending on it. The release action works, but it's just always going to compare against an outdated tag

**Comment by damemi** [comment]
I'm going to keep playing with it in my fork to see if I can figure out some of these problems.

Anyone else should also be able to test it in their forks too, which you should have write access too (since it's your fork).

**Comment by damemi** [comment]
I was wrong about the tag being alphabetical, I think `0.9.0` is just the most recent tag from the master branch (we've been tagging other releases from their release- branch). See this run where I added a new tag to my release-1.18 branch and it checked against that: https://github.com/damemi/descheduler/runs/899793702?check_suite_focus=true

**Comment by damemi** [comment]
Downside to running against master: every merged PR (without a helm version bump) will trigger a chart-releaser action that fails (https://github.com/damemi/descheduler/runs/899803075?check_suite_focus=true#step:4:23). As it should, because we'll only be bumping the chart version during a release (or a backport).

So I think we need to 1) change the action to act on pushes to `release-*` branches and 2) change our normal (non-release) flow to the following:
1. tag the new release from **`master`**
2. cut the `release-1.x` branch

We need to tag before we cut the branch because if not, just cutting the `release-1.x` branch will trigger a new action. Since the last tag on master (and thus the release-1.x branch) is `0.9.0`, it will compare against that and always succeed.

This flow also means that we can merge the helm chart version bump to master in a normal PR as part of our release process.

@seanmalloy @stevehipwell what do you think of this? It's a lot simpler than the flow we originally imagined and still keeps the separate tags for charts (these are created by the action). I updated my pr in https://github.com/kubernetes-sigs/descheduler/pull/356 with these changes

**Comment by stevehipwell** [comment]
@damemi that sounds like a sensible solution if it works with the project flow. You could take a look at how https://github.com/fluxcd/flux do theirs. They use a different action, but it also includes the lint process.

**Comment by damemi** [comment]
> @damemi that sounds like a sensible solution if it works with the project flow. You could take a look at how https://github.com/fluxcd/flux do theirs. They use a different action, but it also includes the lint process.

Flux seems to be using a different chart releaser from `stefanprodan/helm-gh-pages` (https://github.com/fluxcd/flux/blob/master/.github/workflows/chart-release.yml#L12), as opposed to the official `helm/chart-releaser-action` we're trying to use here. I'm assuming that action takes some different steps that don't require the assumptions of the helm one here

**Comment by stevehipwell** [comment]
@damemi yes, the flux flow is the same as the initial one we planned. It looks like the action they're using checks the version in Chart.yaml against the repo metadata on the _gh-pages_ branch when triggered by a `chart-*` tag.

**Comment by seanmalloy** [comment]
Looks like @damemi got us closer ...
* https://github.com/kubernetes-sigs/descheduler/releases/tag/descheduler-helm-chart-0.18.1
* https://github.com/kubernetes-sigs/descheduler/releases/download/descheduler-helm-chart-0.18.1/descheduler-helm-chart-0.18.1.tgz

**Comment by damemi** [comment]
@stevehipwell @seanmalloy Success! With the changes in #357, we have:
* A [successful run of the chart-releaser-action](https://github.com/kubernetes-sigs/descheduler/runs/903881142)
* A [published helm chart release](https://github.com/kubernetes-sigs/descheduler/releases/tag/descheduler-helm-chart-0.18.1)
* An [updated index.yaml](https://github.com/kubernetes-sigs/descheduler/blob/gh-pages/index.yaml) pushed to the `gh-pages` branch (this required removing the CLA check from this branch, or the push failed from the action. Since any changes that could trigger this will need to be first merged to another branch that does require the CLA, I think this is ok. No PRs should ever merge without the CLA, and we shouldn't allow any to be opened directly against this branch anyway).

**Comment by seanmalloy** [comment]
Cool ... thanks for your help @damemi.

I think the final step is to verify that the install instructions work now:
* https://github.com/kubernetes-sigs/descheduler/tree/descheduler-helm-chart-0.18.1/charts/descheduler#tldr
* https://github.com/kubernetes-sigs/descheduler/tree/descheduler-helm-chart-0.18.1/charts/descheduler#installing-the-chart

I think we could also remove all of the files from the `gh-pages` branch except for the file index.yaml, but I guess this optional. After that we can close this issue!

**Comment by damemi** [comment]
@seanmalloy I was able to install the chart using:

```
$ helm repo add descheduler https://kubernetes-sigs.github.io/descheduler/
"descheduler" has been added to your repositorie

$ helm repo update
Hang tight while we grab the latest from your chart repositories...
...Successfully got an update from the "descheduler" chart repository
Update Complete. ⎈ Happy Helming!⎈ 

$ helm install descheduler descheduler/descheduler-helm-chart
NAME: descheduler
LAST DEPLOYED: Thu Jul 23 15:39:32 2020
NAMESPACE: default
STATUS: deployed
REVISION: 1
TEST SUITE: None
NOTES:
Descheduler installed as a cron job.
```

Note that because we changed the chart name from `descheduler` to `descheduler-helm-chart` I had to use that instead of what's in https://github.com/kubernetes-sigs/descheduler/tree/descheduler-helm-chart-0.18.1/charts/descheduler#tldr, so I opened https://github.com/kubernetes-sigs/descheduler/pull/358 to fix that.

Also, really thanks again @stevehipwell for working with us on this and for helping to debug it. It's great to see this working and your and @seanmalloy's efforts to iron out the kinks with me were crucial. It's great to see it working! 🎉

**Comment by jyaworski** [comment]
Is it not possible to update the action to publish the correct name? `helm install descheduler descheduler/descheduler-helm-chart` seems quite strange.

**Comment by damemi** [comment]
> Is it not possible to update the action to publish the correct name? `helm install descheduler descheduler/descheduler-helm-chart` seems quite strange.

We switched the name to avoid confusion between the descheduler releases and the chart releases (which are separate but both hosted in this repo, and they'd both be called "descheduler"). I couldn't find any way to have the action publish the release under a different name than the helm chart name but if anyone else does please feel free to make that change :)

**Comment by stevehipwell** [comment]
Great job @seanmalloy  & @damemi, it was a pleasure working with you!!

**Comment by seanmalloy** [comment]
Final PR to adjust some documentation ... https://github.com/kubernetes-sigs/descheduler/pull/359. I will close this issue once that PR is merged.

**Comment by seanmalloy** [comment]
/close

Thanks @stevehipwell and @damemi for your help with the helm chart. The helm chart is available and all the documentation has been updated.

**Comment by stevehipwell** [comment]
@seanmalloy have you thought about adding the repo to https://hub.helm.sh/?

**Comment by seanmalloy** [comment]
> @seanmalloy have you thought about adding the repo to https://hub.helm.sh/?

I think it is fine if we add the descheduler helm chat there. What is the process to add it? If it's a pull request would you be willing to submit the pull request?

**Comment by stevehipwell** [comment]
@seanmalloy I'm happy to make the PR, I'll just need to know what the contact details should be for the repo (as [required](https://github.com/helm/hub/blob/master/Repositories.md)). You will also want to make sure that the maintainer section of the _Chart.yaml_ file is correct.

**Comment by seanmalloy** [comment]
> @seanmalloy I'm happy to make the PR, I'll just need to know what the contact details should be for the repo (as [required](https://github.com/helm/hub/blob/master/Repositories.md)). You will also want to make sure that the maintainer section of the _Chart.yaml_ file is correct.

@stevehipwell I think we should use email address `kubernetes-sig-scheduling@googlegroups.com` for the contact information. This is the mailing list for SIG Scheduling.

**Comment by stevehipwell** [comment]
FYI @seanmalloy https://github.com/helm/hub/pull/435
